File: CosineFunction.cpp

/*****************************************************************************
 *                                                                           *
 *  CosineFunction.cpp                                                       *
 *  24 September 2006                                                        *
 *  Project: Solus, Ligra                                                    *
 *  Written by: Richard Sartor                                               *
 *  Copyright © 2006 Metaphysics Industries                                  *
 *                                                                           *
 *  The class for the built-in Cosine function.                              *
 *                                                                           *
 *****************************************************************************/



#include "stdafx.h"
#include "CosineFunction.h"
#include "Literal.h"



;
NAMESPACE_START
;



CosineFunction::CosineFunction(void)
{
	this->Name = "Cosine";
}

CosineFunction::~CosineFunction(void)
{
	
}

Literal^ CosineFunction::InternalCall( ... array<Expression^>^ args)
{
	return gcnew Literal((float)Math::Cos(args[0]->Eval()->Value));
}


;
NAMESPACE_END
;

File: Functions/CosineFunction.cs

/*
 *  MetaphysicsIndustries.Solus
 *  Copyright (C) 2006-2022 Metaphysics Industries, Inc., Richard Sartor
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 3 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 *  USA
 *
 */

/*****************************************************************************
 *                                                                           *
 *  CosineFunction.cs                                                        *
 *                                                                           *
 *  The class for the built-in Cosine function.                              *
 *                                                                           *
 *****************************************************************************/

using System;
using System.Collections.Generic;
using MetaphysicsIndustries.Solus.Sets;
using MetaphysicsIndustries.Solus.Values;

namespace MetaphysicsIndustries.Solus.Functions
{
    public class CosineFunction : SingleArgumentFunction
	{
        public static readonly CosineFunction Value = new CosineFunction();

        protected CosineFunction()
		{
		}

        public override string Name => "Cosine";

        public override string DisplayName => "cos";

        public override string DocString
        {
            get
            {
                return "The cosine function\n  cos(x)\n\nReturns the cosine of x.";
            }
        }

        public override ISet GetResultType(SolusEnvironment env,
            IEnumerable<ISet> argTypes) => Reals.Value;
    }
}
